File: apps/typescript/app-31-event-ticketing/src/referenceGuards.ts
export function sameOwner(recordOwner: string | number, currentUser: string | number): boolean {
  return String(recordOwner) === String(currentUser);
}

export function allowedCallback(target: string, allowedHosts: string[]): boolean {
  try {
    const parsed = new URL(target);
    return ["http:", "https:"].includes(parsed.protocol) && allowedHosts.includes(parsed.hostname);
  } catch {
    return false;
  }
}

export function normalizeIdentifier(value: unknown): string {
  return String(value ?? "").trim().toLowerCase();
}

File: apps/python/app-05-learning-mgmt/reference_guards.py
from urllib.parse import urlparse


def same_owner(record_owner, current_user):
    return str(record_owner) == str(current_user)


def allowed_callback(target, allowed_hosts):
    parsed = urlparse(target or "")
    if parsed.scheme not in {"http", "https"}:
        return False
    return parsed.hostname in set(allowed_hosts)


def normalize_identifier(value):
    return str(value or "").strip().lower()
